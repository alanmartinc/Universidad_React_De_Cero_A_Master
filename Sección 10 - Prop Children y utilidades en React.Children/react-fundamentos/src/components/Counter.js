import React, {useState} from 'react'

export const Button = ({type, onIncrement, onDecrement}) => {

  const action = () => {
    return type === 'increment'
      ? onIncrement()
      : onDecrement()
  }

  return (
    <button onClick={action}>
      {type === 'increment' ? 'Agregar' : 'Quitar'}
    </button>
  )
}

export const Title = ({clicks, children}) => {

  return typeof children === 'function'
    ? children(clicks)
    : <span>{ clicks }</span>
}

export const Counter = ({children}) => {
  const [clicks, setClicks]= useState(0)

  const increment = () => setClicks(clicks + 1)
  const decrement = () => setClicks(clicks - 1)

  if (!children) {
    const stlyes = {
      background: 'tomato',
      borderRadius: '0.2em',
      padding: '0.3em 1em',
      color: '#FFF'
    }

    return (
      <div style={stlyes}>
        Wops! Debes agregar componentes como {'<Button />'}
      </div>
    )
  }

  const _children = React.Children.map(children, (child) => {
    console.log(child)

    let props = {}

    if (child.type === Title) {
      props.clicks = clicks
    }

    if (child.type === Button) {
      props.onIncrement = increment
      props.onDecrement = decrement
    }

    return React.cloneElement(child, props)
  })

  return _children
}
