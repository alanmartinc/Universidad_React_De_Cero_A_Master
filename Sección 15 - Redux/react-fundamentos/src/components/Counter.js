import React from 'react'
import { connect } from 'react-redux'
import { increment, decrement } from '../redux/actions/counterActions'

const Counter = ({ increment, decrement, counter, name }) => {
  return (
    <div>
      <button onClick={increment}>+</button>
      <button onClick={decrement}>-</button>
      <h1>{ counter }</h1>
      <p>
        { name }
      </p>
    </div>
  )
}

const mapStateToProps = (state) => {
  return {
    name: state.user.name,
    counter: state.counter
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    increment: () => dispatch(increment()),
    decrement: () => dispatch(decrement())
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Counter)