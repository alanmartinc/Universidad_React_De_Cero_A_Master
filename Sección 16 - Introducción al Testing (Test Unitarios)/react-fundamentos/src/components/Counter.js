import React, {useState} from 'react'

const Counter = () => {
    const [clicks, setClicks] = useState(0)

    const increment = () => setClicks(clicks + 1)
    const decrement = () => setClicks(clicks - 1)

    return(
        <div>
            <button onClick={increment}>
                Increment
            </button>

            <button onClick={decrement}>
                Decrement
            </button>
            <h1>{clicks}</h1>
        </div>
    )
}

export default Counter