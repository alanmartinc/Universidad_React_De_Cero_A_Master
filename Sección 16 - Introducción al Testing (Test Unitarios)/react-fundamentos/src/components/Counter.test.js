import {configure, shallow, mount, render} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import Counter from './Counter'

configure({
    adapter: new Adapter()
})

describe('Probando componente <Counter/>', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallow(<Counter/>)
    })

    test('Validar que coincida con Snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    test('Validar funcionamiento de botones', () => {
        wrapper.find('button').first().simulate('click')

        expect(wrapper.find('h1').text()).toBe('1')

        wrapper.find('button').last().simulate('click')
        wrapper.find('button').last().simulate('click')
        wrapper.find('button').last().simulate('click')

        expect(wrapper.find('h1').text()).toBe('-2')
    })
})