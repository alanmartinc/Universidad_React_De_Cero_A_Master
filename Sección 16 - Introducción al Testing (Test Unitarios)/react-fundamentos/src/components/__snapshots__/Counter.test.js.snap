// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando componente <Counter/> Validar que coincida con Snapshot 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Increment
  </button>
  <button
    onClick={[Function]}
  >
    Decrement
  </button>
  <h1>
    0
  </h1>
</div>
`;
